harness/determined/profiler.py

```diff
@@ -8,17 +14,61 @@
 import psutil
 import pynvml
 
+from timeit import default_timer
+import determined as det
 from determined.common import api
 from determined.common.api import TrialProfilerMetricsBatch
 
 SYSTEM_METRIC_TYPE_ENUM = "PROFILER_METRIC_TYPE_SYSTEM"
-
+TIMING_METRIC_TYPE_ENUM = "PROFILER_METRIC_TYPE_TIMING"
 LOG_NAMESPACE = "determined-profiler"
 
 
+def build_profiler_agent(env: det.EnvContext) -> 'ProfilerAgent':
+    if env.experiment_config.profiling_enabled() and get_rank_if_horovod_process_else_return_zero() == 0:
+        begin_on_batch, end_after_batch = env.experiment_config.profiling_interval()
+        return SystemProfilerAgent(
+            env.det_trial_id,
+            env.det_agent_id,
+            env.master_url,
+            begin_on_batch,
+            end_after_batch
+        )
+    return NoopProfilerAgent()
 
 
-class ProfilerAgent:
+class ProfilerAgent(metaclass=abc.ABCMeta):
+    """
+    Base agent when profiling is disabled.
+    """
+    @abc.abstractmethod
+    def update_batch_idx(self, new_batch_idx: int):
+        pass
+
+    @abc.abstractmethod
+    def record_timing(self, metric_name: str):
+        pass
+
+
+class NoopProfilerAgent(ProfilerAgent):
```

Comment:
I don't think we need a secondary class. The ProfilerAgent has a property called `is_enabled`, and all of the operations should be no-ops if that property is false. We will need that logic in order to keep the API consistent after the profiling is done, so I think we should reuse it instead of adding NoopProfilerAgent.

Reply:
I have an abstract class to keep the interface in line. I just thought in real code, it was better to pass then accidentally do anything in real training code.